src/os/hostname.zig

```diff
@@ -1,11 +1,102 @@
 const std = @import("std");
+const builtin = @import("builtin");
 const posix = std.posix;
 
 pub const HostnameParsingError = error{
     NoHostnameInUri,
     NoSpaceLeft,
 };
 
+pub const UrlParsingError = std.Uri.ParseError || error{
+    HostnameIsNotMacAddress,
+    NoSchemeProvided,
+};
+
+const mac_address_length = 17;
+
+fn isUriPathSeparator(c: u8) bool {
+    return switch (c) {
+        '?', '#' => true,
+        else => false,
+    };
+}
+
+fn isValidMacAddress(mac_address: []const u8) bool {
+    // A valid mac address has 6 two-character components with 5 colons, e.g. 12:34:56:ab:cd:ef.
+    if (mac_address.len != 17) {
+        return false;
+    }
+
+    for (mac_address, 0..) |c, i| {
+        if ((i + 1) % 3 == 0) {
+            if (c != ':') {
+                return false;
+            }
+        } else if (!std.mem.containsAtLeastScalar(u8, "0123456789ABCDEFabcdef", 1, c)) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+/// Parses the provided url to a `std.Uri` struct. This is very specific to getting hostname and
+/// path information for Ghostty's PWD reporting functionality. Takes into account that on macOS
+/// the url passed to this function might have a mac address as its hostname and parses it
+/// correctly.
+pub fn parseUrl(url: []const u8) UrlParsingError!std.Uri {
+    return std.Uri.parse(url) catch |e| {
+        // The mac-address-as-hostname issue is specific to macOS so we just return an error if we
+        // hit it on other platforms.
+        if (comptime builtin.os.tag != .macos) return e;
```

Comment:
@jparise I changed this from `comptime if (…` to `if (comptime …` to fix the build failure on non macOS systems, but I'm not familiar enough with Zig's comptime to know if it makes sense to even keep the `comptime` check in here? Should we just remove it, or does it still make sense to keep it? 🤔

Reply:
In its current state, I don't think we need to gate this on macOS. It's done in a generic way, although it would be good to get some real-world confirmation that this all makes sense on other platforms (e.g. no false positives/negatives).

Reply:
@jcollie and @pluiedev, do you have any opinions on how this will play on Linux?